crates/symbolicator/src/services/download/mod.rs

```diff
@@ -199,23 +200,142 @@ async fn download_stream(
     stream: impl Stream<Item = Result<impl AsRef<[u8]>, DownloadError>>,
     destination: PathBuf,
 ) -> Result<DownloadStatus, DownloadError> {
+    let source = source.into();
+
     // All file I/O in this function is blocking!
-    log::trace!("Downloading from {}", source.into());
+    log::trace!("Downloading from {}", source);
     let mut file = File::create(&destination)
         .await
         .map_err(DownloadError::BadDestination)?;
     futures::pin_mut!(stream);
 
-    while let Some(chunk) = stream.next().await {
-        let chunk = chunk?;
-        file.write_all(chunk.as_ref())
-            .await
-            .map_err(DownloadError::Write)?;
+    let mut throughput_recorder =
+        MeasureSourceDownloadGuard::new("source.stream", source.source_metric_key());
+    let result: Result<_, DownloadError> = async {
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk?;
+            let chunk = chunk.as_ref();
+            throughput_recorder.add_bytes_transferred(chunk.len() as u64);
+            file.write_all(chunk).await.map_err(DownloadError::Write)?;
+        }
+        Ok(())
     }
+    .await;
+    throughput_recorder.done(&result);
+    result?;
+
     file.flush().await.map_err(DownloadError::Write)?;
     Ok(DownloadStatus::Completed)
 }
 
+/// State of the [`MeasureSourceDownloadGuard`].
+#[derive(Clone, Copy, Debug)]
+enum MeasureState {
+    /// The future is not ready.
+    Pending,
+    /// The future has terminated with a status.
+    Done(&'static str),
+}
+
+/// A guard to [`measure`] the amount of time it takes to download a source. This guard is also
+/// capable of calculating and reporting the throughput of the connection. Two metrics are
+/// emitted if `bytes_transferred` is set:
+///
+/// 1. Amount of time taken to complete the measurement
+/// 2. Connection thoroughput (bytes transferred / time taken to complete)
+///
+/// If `bytes_transferred` is not set, then only the first metric (amount of time taken) is
+/// recorded.
+struct MeasureSourceDownloadGuard<'a> {
+    state: MeasureState,
+    task_name: &'a str,
+    source_name: &'a str,
+    creation_time: Instant,
+    bytes_transferred: Option<u64>,
+}
+
+impl<'a> MeasureSourceDownloadGuard<'a> {
+    /// Creates a new measure guard for downloading a source.
+    pub fn new(task_name: &'a str, source_name: &'a str) -> Self {
+        Self {
+            state: MeasureState::Pending,
+            task_name,
+            source_name,
+            bytes_transferred: None,
+            creation_time: Instant::now(),
+        }
+    }
+
+    /// A checked add to the amount of bytes transferred during the download.
+    ///
+    /// This value will be emitted when the download's future is completed or cancelled.
+    pub fn add_bytes_transferred(&mut self, additional_bytes: u64) {
+        let bytes = self.bytes_transferred.get_or_insert(Default::default());
+        *bytes = bytes.saturating_add(additional_bytes);
```

Comment:
Might as well use `get_or_insert(0)`, since its a numeric value.

Since this is bytes that we add, we will overflow storage sooner than wrapping around a `u64`. But the `saturating_add` is a lot easier to read than the `checked_add` you had before, that was my main concern.